# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: netwerk/cookie/CookieService.cpp
# Commit: 2a8ebc224db0
# Full Hash: 2a8ebc224db092bc92e670dc5de18b5bf3303685
# Author: Tim Huang <tihuang@mozilla.com>
# Date: 2024-12-20 21:45:56
# Description:
#   Bug 1938478 - Don't check third-party cookie blocking exceptions after the shutdown is confirmed. r=valentin,cookie-reviewers
#   
#   Getting the browsing context in a MozPromise callback during shutdown
#   can cause crash issues. To prevent this, we don't check third-party
#   cookie blocking exceptions after the shutdown is confirmed because it
# ==============================================================================

diff -r b66436b1a867 -r 2a8ebc224db0 netwerk/cookie/CookieService.cpp
--- a/netwerk/cookie/CookieService.cpp	Fri Dec 20 11:23:02 2024 +0000
+++ b/netwerk/cookie/CookieService.cpp	Fri Dec 20 11:27:28 2024 +0000
@@ -1769,6 +1769,12 @@
       GetMainThreadSerialEventTarget(), __func__,
       [channel = nsCOMPtr{aChannel}, csSingleton, loadInfo](
           const GenericNonExclusivePromise::ResolveOrRejectValue& aValue) {
+        // We don't check the 3PCB exception list if the app is shutting down.
+        if (AppShutdown::IsInOrBeyond(ShutdownPhase::AppShutdownConfirmed)) {
+          channel->Resume();
+          return NS_OK;
+        }
+
         // We check the 3PCB exception list here. We will check both the
         // wildcard exception and the specific exception. If any of them is in
         // the exception list, we will set the channel's isOn3PCBExceptionList
